hail/src/main/scala/is/hail/expr/types/physical/PArray.scala

```diff
@@ -1,43 +1,21 @@
 package is.hail.expr.types.physical
 
-import is.hail.annotations.{UnsafeUtils, _}
-import is.hail.asm4s._
-import is.hail.check.Gen
+import is.hail.annotations.CodeOrdering
 import is.hail.expr.ir.EmitMethodBuilder
 import is.hail.expr.types.virtual.TArray
-import is.hail.utils._
-import scala.reflect.{ClassTag, _}
+object PArray {
+  def apply(elementType: PType, required: Boolean = false) = new PCanonicalArray(elementType, required)
+}
 
-final case class PArray(elementType: PType, override val required: Boolean = false) extends PContainer with PStreamable {
+abstract class PArray extends PContainer with PStreamable {
   lazy val virtualType: TArray = TArray(elementType.virtualType, required)
 
-  val elementByteSize: Long = UnsafeUtils.arrayElementSize(elementType)
-
-  val contentsAlignment: Long = elementType.alignment.max(4)
-
-  override def pyString(sb: StringBuilder): Unit = {
-    sb.append("array<")
-    elementType.pyString(sb)
-    sb.append('>')
-  }
-  override val fundamentalType: PArray = {
-    if (elementType == elementType.fundamentalType)
-      this
-    else
-      this.copy(elementType = elementType.fundamentalType)
-  }
-
-  def _asIdent = s"array_of_${elementType.asIdent}"
-  def _toPretty = s"Array[$elementType]"
-
-  override def _pretty(sb: StringBuilder, indent: Int, compact: Boolean = false) {
-    sb.append("Array[")
-    elementType.pretty(sb, indent, compact)
-    sb.append("]")
-  }
+  def copy(elementType: PType = this.elementType, required: Boolean = this.required): PArray
```

Comment:
We should probably get rid of this when possible.

Reply:
I can get rid of this here. It's only used in PType, and is trivial I think.

Reply:
Actually, I think the solution I was thinking of `t.copy -> PArray(` may not work well (since the apply method on PArray would need to know what to do based on the input type, which is a copy method anyway). The other option I can see is to allow requiredeness on the PType to be modified / be a var.

For reference this is only used in

```scala
final def setRequired(required: Boolean): PType = {
    if (required == this.required)
      this
    else
      this match {
        case PBinary(_) => PBinary(required)
        case PBoolean(_) => PBoolean(required)
        case PInt32(_) => PInt32(required)
        case PInt64(_) => PInt64(required)
        case PFloat32(_) => PFloat32(required)
        case PFloat64(_) => PFloat64(required)
        case PString(_) => PString(required)
        case PCall(_) => PCall(required)
        case t: PArray => t.copy(required = required)
        case t: PSet => t.copy(required = required)
        case t: PDict => t.copy(required = required)
        case t: PLocus => t.copy(required = required)
        case t: PInterval => t.copy(required = required)
        case t: PStruct => t.copy(required = required)
        case t: PTuple => t.copy(required = required)
      }
  }
```

Reply:
yeah don't change that here. Violation of the principle that *reorganizational changes* (moving large blocks of code as unchanged as possible tween classes/files) and semantic changes (changes to the meaning of any implementation) shouldn't be mixed, because the presence of any semantic changes means that hundreds of lines of simple organizational changes need to be looked at closely for interaction effects.